File: sidekick/App/App/Plugins/CacophonyPlugin.swift
//
//  DashboardPlugin.swift
//  App
//
//  Created by Patrick Baxter on 29/11/22.
//

import Capacitor
import shared
var documentPath = String(FileManager.default.urls(for: .documentDirectory, in: .userDomainMask)[0].absoluteString.dropFirst(7));

@objc(CacophonyPlugin)
public class CacophonyPlugin: CAPPlugin, CAPBridgedPlugin {
    public var identifier: String = "CacophonyPlugin"
    public var jsName: String = "Cacophony"
    
    public let pluginMethods: [CAPPluginMethod] = [
    CAPPluginMethod(name:"authenticateUser", returnType: CAPPluginReturnPromise),
    CAPPluginMethod(name:"requestDeletion", returnType: CAPPluginReturnPromise),
    CAPPluginMethod(name:"validateToken", returnType: CAPPluginReturnPromise),
    CAPPluginMethod(name:"setToTestServer", returnType: CAPPluginReturnPromise),
    CAPPluginMethod(name:"setToProductionServer", returnType: CAPPluginReturnPromise),
    CAPPluginMethod(name:"setToCustomServer", returnType: CAPPluginReturnPromise),
    CAPPluginMethod(name:"uploadRecording", returnType: CAPPluginReturnPromise),
    CAPPluginMethod(name:"getStationsForUser", returnType: CAPPluginReturnPromise),
    CAPPluginMethod(name:"updateStation", returnType: CAPPluginReturnPromise),
    CAPPluginMethod(name:"getReferenceImage", returnType: CAPPluginReturnPromise),
    CAPPluginMethod(name:"getReferencePhoto", returnType: CAPPluginReturnPromise),
    CAPPluginMethod(name:"uploadReferencePhoto", returnType: CAPPluginReturnPromise),
    CAPPluginMethod(name:"deleteReferencePhoto", returnType: CAPPluginReturnPromise),
    CAPPluginMethod(name:"createStation", returnType: CAPPluginReturnPromise),
    CAPPluginMethod(name:"uploadEvent", returnType: CAPPluginReturnPromise),
    CAPPluginMethod(name:"getAppVersion", returnType: CAPPluginReturnPromise),
    ]
    @objc let cacophony = CacophonyInterface(filePath: documentPath)

    @objc func authenticateUser(_ call: CAPPluginCall) {
        DispatchQueue.global().async { [weak self] in
            self?.cacophony.authenticateUser(call: pluginCall(call: call))
        }
    }
    @objc func requestDeletion(_ call: CAPPluginCall) {
        DispatchQueue.global().async { [weak self] in
            self?.cacophony.requestDeletion(call: pluginCall(call: call))
        }
    }
    @objc func validateToken(_ call: CAPPluginCall) {
        DispatchQueue.global().async { [weak self] in
            self?.cacophony.validateToken(call: pluginCall(call: call))
        }
    }
    @objc func setToTestServer(_ call: CAPPluginCall) {
        DispatchQueue.global().async { [weak self] in
            self?.cacophony.setToTestServer(call: pluginCall(call: call))
        }
    }
    @objc func setToProductionServer(_ call: CAPPluginCall) {
        DispatchQueue.global().async { [weak self] in
            self?.cacophony.setToProductionServer(call: pluginCall(call: call))
        }
    }
    @objc func setToCustomServer(_ call: CAPPluginCall) {
        DispatchQueue.global().async { [weak self] in
            self?.cacophony.setToCustomServer(call: pluginCall(call: call))
        }
    }
    @objc func uploadRecording(_ call: CAPPluginCall) {
        DispatchQueue.global().async { [weak self] in
            self?.cacophony.uploadRecording(call: pluginCall(call: call))
        }
    }
    @objc func uploadEvent(_ call:CAPPluginCall) {
        DispatchQueue.global().async { [weak self] in
            self?.cacophony.uploadEvent(call: pluginCall(call: call))
        }
    }
    @objc func getStationsForUser(_ call:CAPPluginCall) {
        DispatchQueue.global().async { [weak self] in
            self?.cacophony.getStationsForUser(call: pluginCall(call: call))
        }
    }
    @objc func updateStation(_ call:CAPPluginCall) {
        DispatchQueue.global().async { [weak self] in
            self?.cacophony.updateStation(call: pluginCall(call: call))
        }
    }
    @objc func uploadReferencePhoto(_ call:CAPPluginCall) {
        DispatchQueue.global().async { [weak self] in
            self?.cacophony.uploadReferencePhoto(call: pluginCall(call: call))
        }
    }
    @objc func getReferenceImage(_ call:CAPPluginCall) {
        DispatchQueue.global().async { [weak self] in
            self?.cacophony.getReferenceImage(call: pluginCall(call: call))
        }
    }
    @objc func getReferencePhoto(_ call:CAPPluginCall) {
        DispatchQueue.global().async { [weak self] in
            self?.cacophony.getReferencePhoto(call: pluginCall(call: call))
        }
    }
    @objc func deleteReferencePhoto(_ call:CAPPluginCall) {
        DispatchQueue.global().async { [weak self] in
            self?.cacophony.deleteReferencePhoto(call: pluginCall(call: call))
        }
    }
    @objc func createStation(_ call:CAPPluginCall) {
        DispatchQueue.global().async { [weak self] in
            self?.cacophony.createStation(call: pluginCall(call: call))
        }
    }
    @objc func getAppVersion(_ call: CAPPluginCall) {
        DispatchQueue.global().async { 
            if let version = Bundle.main.infoDictionary?["CFBundleShortVersionString"] as? String,
               let patchVersion = Bundle.main.infoDictionary?["CFBundleVersion"] as? String {
                let fullVersion = "\(version).\(patchVersion)"
                call.resolve(["data": fullVersion, "success": true])
            } else {
                call.resolve(["data": "1.0.0", "success": true])
            }
        }
    }
}

File: sidekick/shared/src/androidMain/kotlin/nz/org/cacophony/sidekick/CacophonyPlugin.kt
package nz.org.cacophony.sidekick

import android.content.pm.PackageInfo
import android.content.pm.PackageManager
import android.os.Build
import com.getcapacitor.JSObject
import com.getcapacitor.Plugin
import com.getcapacitor.PluginCall
import com.getcapacitor.PluginMethod
import com.getcapacitor.annotation.CapacitorPlugin
import nz.org.cacophony.sidekick.cacophony.CacophonyInterface

@CapacitorPlugin(name = "Cacophony")
class CacophonyPlugin : Plugin() {
    lateinit var cacophony: CacophonyInterface

    override fun load() {
        cacophony = CacophonyInterface(context.applicationContext.filesDir.absolutePath)
    }

    @PluginMethod
    fun authenticateUser(call: PluginCall) {
        cacophony.authenticateUser(pluginCall(call))
    }

    @PluginMethod
    fun requestDeletion(call: PluginCall) {
        cacophony.requestDeletion(pluginCall(call))
    }

    @PluginMethod
    fun validateToken(call: PluginCall) {
        cacophony.validateToken(pluginCall(call))
    }

    @PluginMethod
    fun setToTestServer(call: PluginCall) {
        cacophony.setToTestServer(pluginCall(call))
    }

    @PluginMethod
    fun setToProductionServer(call: PluginCall) {
        cacophony.setToProductionServer(pluginCall(call))
    }

    @PluginMethod
    fun setToCustomServer(call: PluginCall) {
        cacophony.setToCustomServer(pluginCall(call))
    }

    @PluginMethod
    fun uploadRecording(call: PluginCall) {
        cacophony.uploadRecording(pluginCall(call))
    }

    @PluginMethod
    fun uploadEvent(call: PluginCall) {
        cacophony.uploadEvent(pluginCall(call))
    }

    @PluginMethod
    fun getDeviceById(call: PluginCall) {
        cacophony.getDeviceById(pluginCall(call))
    }

    @PluginMethod
    fun getStationsForUser(call: PluginCall) {
        cacophony.getStationsForUser(pluginCall(call))
    }

    @PluginMethod
    fun createStation(call: PluginCall) {
        cacophony.createStation(pluginCall(call))
    }

    @PluginMethod
    fun updateStation(call: PluginCall) {
        cacophony.updateStation(pluginCall(call))
    }

    @PluginMethod
    fun uploadReferencePhoto(call: PluginCall) {
        cacophony.uploadReferencePhoto(pluginCall(call))
    }

    @PluginMethod
    fun getReferencePhoto(call: PluginCall) {
        cacophony.getReferencePhoto(pluginCall(call))
    }

    @PluginMethod
    fun getReferenceImage(call: PluginCall) {
        cacophony.getReferenceImage(pluginCall(call))
    }


    @PluginMethod
    fun saveDeviceImage(call: PluginCall) {
        cacophony.saveDeviceImage(pluginCall(call))
    }

    @PluginMethod
    fun uploadDeviceReferenceImage(call: PluginCall) {
        cacophony.uploadDeviceReferenceImage(pluginCall(call))
    }

    @PluginMethod
    fun deleteReferenceImage(call: PluginCall) {
        cacophony.deleteReferenceImage(pluginCall(call))
    }


    @PluginMethod
    fun deleteReferencePhoto(call: PluginCall) {
        cacophony.deleteReferencePhoto(pluginCall(call))
    }


    fun PackageManager.getPackageInfoCompat(packageName: String, flags: Int = 0): PackageInfo =
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU) {
            getPackageInfo(packageName, PackageManager.PackageInfoFlags.of(flags.toLong()))
        } else {
            @Suppress("DEPRECATION") getPackageInfo(packageName, flags)
        }

    @PluginMethod
    fun getAppVersion(call: PluginCall) {
        // Get version name
        try {
            val pInfo = context.packageManager.getPackageInfoCompat(context.packageName)
            val version = pInfo.versionName
            val data = JSObject()
            data.put("data", version)
            data.put("success", true)
            call.resolve(data)
        } catch (e: PackageManager.NameNotFoundException) {
            e.printStackTrace()
            val data = JSObject()
            data.put("success", false)
            data.put("message", "Could not get app version")
            call.resolve(data)
        }

    }
}

